src/mutex.c

```diff
@@ -65,6 +65,49 @@ JEMALLOC_EXPORT int	_pthread_mutex_init_calloc_cb(pthread_mutex_t *mutex,
     void *(calloc_cb)(size_t, size_t));
 #endif
 
+void
+malloc_mutex_lock_slow(malloc_mutex_t *mutex) {
+	lock_prof_data_t *data = &mutex->prof_data;
+	bool spin_success = false;
+
+	{//TODO: a smart spin policy
+		if (!malloc_mutex_trylock(mutex)) {
+			spin_success = true;
+			goto label_locked;
+		}
+	}
+
+	nstime_t now, before;
+	uint32_t n_thds;
+	nstime_init(&now, 0);
+	nstime_update(&now);
+	n_thds = atomic_add_u32(&data->n_waiting_thds, 1);
+	/* One last try as above two calls may take quite some cycles. */
+	if (!malloc_mutex_trylock(mutex)) {
+		spin_success = true;
+		atomic_sub_u32(&data->n_waiting_thds, 1);
+		goto label_locked;
+	}
+	nstime_copy(&before, &now);
```

Comment:
Oh now I remember why I kept "now" to be mutated -- it's to avoid the clock skew by non-monotonic clocks. For example, if we do:
```
nstime_t before = 0, after = 0;
nstime_update(&before);
LOCK_STUFF;
nstime_update(&after);
nstime_copy(&delta, &after);
```
Then there's no guarantee that now >= before, as nstime_update only checks if the current clock is moving forward. So I decided on having a "now" and only update on that.

But I agree it should be made clearer about the naming and non-monotonic issue. How about 
```
nstime_t after, before = NSTIME_ZERO_INITIALIZER;
nstime_update(&before);
/* Copy before to after to make avoid skew caused by non monotonic clocks. */
nstime_copy(&after, &before);
LOCK_STUFF()...
nstime_update(&after);
nstime_copy(&delta, &after);
nstime_substract(&delta, &before);
```